File: frontend/app/js/_test/schemas/comic/index.spec.js
import { receiveAll } from '../../../schemas/comic'
import comicResponse from '../../fixtures/comic'
import paginationResponse from '../../fixtures/pagination'

describe('schemas',() => {
  describe('comic',() => {
    describe('receiveAll',() => {
      it('returns the normalized result', () => {
        const comic = comicResponse()
        const pagination = paginationResponse()
        const response = {
          pagination,
          results: [comic]
        }
        const expected = {
          result: { results: [1], pagination },
          entities: {
            comic: {
              1: comic
            }
          }
        }

        expect(receiveAll(response)).toEqual(expected)
      })
    })
  })
})

File: api/spec/concepts/comic/operation/index_spec.rb
require 'rails_helper'

describe Comic::Index do
  before :each do
    stub_const('API_CLIENT', instance_double(Marvel::Api).as_null_object)
    allow(API_CLIENT).to receive(:search_comics).with(
      characters: 1, page: 0,
    ).and_return(
      status: 200,
      pagination: {
        currentPage: 0,
        totalPages: 0,
        isLast: true,
        totalElements: 0,
      },
      comics: [],
    )
  end

  it 'returns the correct result' do
    result = Comic::Index.(characters: 1, page: 0)

    expect(result['results.comics']).to eql([])
    expect(result['results.pagination']).to eql(
      currentPage: 0,
      totalPages: 0,
      isLast: true,
      totalElements: 0,
    )
  end

  context 'when the request is a failure' do
    before :each do
      allow(API_CLIENT).to receive(:search_comics).with(
        characters: 1, page: 0,
      ).and_return(
        status: 409,
        message: 'Error',
      )
    end

    it 'returns an error code' do
      result = Comic::Index.(characters: 1, page: 0)

      expect(result).to be_failure
    end
  end
end
